packages/server/lib/StudioLifecycleManager.ts

```diff
@@ -0,0 +1,157 @@
+        const script = await api.getCaptureProtocolScript(protocolUrl)
+
+        await protocolManager.prepareProtocol(script, {
+          runId: 'studio',
+          projectId: cfg.projectId,
+          testingType: cfg.testingType,
+          cloudApi: {
+            url: routes.apiUrl,
+            retryWithBackoff: api.retryWithBackoff,
+            requestPromise: api.rp,
+          },
+          projectConfig: _.pick(cfg, ['devServerPublicPathRoute', 'port', 'proxyUrl', 'namespace']),
+          mountVersion: api.runnerCapabilities.protocolMountVersion,
+          debugData,
+          mode: 'studio',
+        })
+
+        studioManager.protocolManager = protocolManager
+      } else {
+        debug('Cloud studio is not enabled - skipping protocol setup')
+      }
+
+      debug('Studio is ready')
+      this.studioManager = studioManager
+      this.callRegisteredListeners()
+
+      return studioManager
+    }).catch(async (error) => {
+      debug('Error during studio manager setup: %o', error)
+
+      const cloudEnv = (process.env.CYPRESS_INTERNAL_ENV || 'production') as 'development' | 'staging' | 'production'
+      const cloudUrl = ctx.cloud.getCloudUrl(cloudEnv)
+      const cloudHeaders = await ctx.cloud.additionalHeaders()
+
+      reportStudioError({
+        cloudApi: {
+          cloudUrl,
+          cloudHeaders,
+          CloudRequest,
+          isRetryableError,
+          asyncRetry,
+        },
+        studioHash: projectId,
+        projectSlug: cfg.projectId,
+        error,
+        studioMethod: 'initializeStudioManager',
+        studioMethodArgs: [],
+      })
+
+      // Clean up any registered listeners
+      this.listeners = []
+
+      return null
+    })
+
+    this.studioManagerPromise = studioManagerPromise
+
+    // Register this instance in the data context
+    ctx.update((data) => {
+      data.studioLifecycleManager = this
+    })
+  }
+
+  isStudioReady (): boolean {
+    return !!this.studioManager
+  }
+
+  async getStudio () {
+    if (!this.studioManagerPromise) {
+      throw new Error('Studio manager has not been initialized')
+    }
+
+    return await this.studioManagerPromise
+  }
+
+  private callRegisteredListeners () {
+    if (!this.studioManager) {
+      throw new Error('Studio manager has not been initialized')
+    }
+
+    const studioManager = this.studioManager
+
+    debug('Calling all studio ready listeners')
+    this.listeners.forEach((listener) => {
+      listener(studioManager)
+    })
+
+    this.listeners = []
+  }
+
+  /**
+   * Register a listener that will be called when the studio is ready
+   * @param listener Function to call when studio is ready
+   */
+  registerStudioReadyListener (listener: (studioManager: StudioManager) => void): void {
+    // if studio is already ready and there is a studio manager, call the listener immediately and only once
+    if (this.isStudioReady() && this.studioManager) {
```

Comment:
I don't think both checks are needed.

```suggestion
    // if there is already a studio manager, call the listener immediately
    if (this.studioManager) {
```